shelf.go

```diff
@@ -80,12 +80,12 @@ func openShelf(path string, slotSize uint32, onData onShelfDataFn, readonly bool
 		fileSize int
 		h        = shelfHeader{Magic, curVersion, slotSize}
 		fname    = fmt.Sprintf("bkt_%08d.bag", slotSize)
-		flags    = os.O_RDWR | os.O_CREATE
 	)
-	if readonly {
-		flags = os.O_RDONLY
-	}
-	f, err := os.OpenFile(filepath.Join(path, fname), flags, 0666)
+	// Max 5 files @ 2GB each.
+	// We also want to ensure that the cap is a multiple of the slot size, so no
+	// slot crosses a file boundary.
+	cap := uint64(2*1024*1024 - 2*1024*1024%slotSize)
```

Comment:
This counter does not take into account the billy magic and shelf header metadata. That would push all these "calculated" offsets to be wrong since the first file would have some extra junk that would shift everything (and would need a different cap for the first and subsequent data files).

Reply:
Ouch! Good catch!

Reply:
I guess one way to handle if differently, without tying capped file too deeply with shelves, is to have the following rule:
- On write: if `offset + len(b) > cap`, then write to the 'next' file. And similarly:
- On read: if `offset + len(b) > cap`, then read from the 'next' file.

In other words: let the max byte determine the position of the read/write. This has the following characteristics

- Reading `1` byte may read from different location than reading `n` bytes and picking out the first. 
- The `cap` will never be exceeded, 
- As long as read and writes are always done chunkwise, everything works fine. Non-chunked read/writes will behave undefined.

Reply:
Ah, scrap that, won't work. The other even more simple alternative is to allow chunks to overflow the file-cap, and not spread the read/write out across several files. Then the limit is more of a recommendation, which is fine in our case (we use 2GB but the 'hard' limit is 4GB so we've got space to spare)